codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/endpoints/EndpointParamsGenerator.kt

```diff
@@ -0,0 +1,282 @@
+                    "param" to writable {
+                        when {
+                            type.isOptional() && type.rustType().isCopy() -> rust("self.$name")
+                            type.isOptional() -> rust("self.$name.as_deref()")
+                            type.rustType().isCopy() -> rust("Some(self.$name)")
+                            else -> rust("Some(&self.$name)")
+                        }
+                    },
+                )
+            }
+        }
+    }
+
+    private fun value(value: Value): String {
+        return when (value) {
+            is Value.Str -> value.value().dq() + ".to_string()"
+            is Value.Bool -> value.expectBool().toString()
+            else -> TODO("unexpected type: $value")
+        }
+    }
+
+    private fun generateEndpointParamsBuilder(rustWriter: RustWriter) {
+        rustWriter.docs("Builder for [`Params`]")
+        Attribute.Derives(setOf(Debug, Default, PartialEq, Clone)).render(rustWriter)
+        rustWriter.rustBlock("pub struct Builder") {
+            endpointRules.parameters.toList().forEach { parameter ->
+                val name = parameter.memberName()
+                val type = parameter.symbol().makeOptional()
+                rust("$name: #T,", type)
+            }
+        }
+
+        rustWriter.rustBlock("impl Builder") {
+            docs("Consume this builder, creating [`Params`].")
+            rustBlockTemplate(
+                "pub fn build(self) -> Result<#{Params}, #{ParamsError}>",
+                "Params" to paramsStruct(),
+                "ParamsError" to paramsError(),
+            ) {
+                val params = writable {
+                    rustBlockTemplate("#{Params}", "Params" to paramsStruct()) {
+                        endpointRules.parameters.toList().forEach { parameter ->
+                            rust("${parameter.memberName()}: self.${parameter.memberName()}")
+                            parameter.default.orNull()?.also { default -> rust(".or(Some(${value(default)}))") }
+                            if (parameter.isRequired) {
+                                rustTemplate(
+                                    ".ok_or_else(||#{Error}::missing(${parameter.memberName().dq()}))?",
+                                    "Error" to paramsError(),
+                                )
+                            }
+                            rust(",")
+                        }
+                    }
+                }
+                rust("Ok(#W)", params)
+            }
+            endpointRules.parameters.toList().forEach { parameter ->
+                val name = parameter.memberName()
+                val type = parameter.symbol().mapRustType { t -> t.stripOuter<RustType.Option>() }
+                rustTemplate(
+                    """
+                    /// Sets the value for $name #{extraDocs:W}
+                    pub fn $name(mut self, value: #{type}) -> Self {
```

Comment:
Should this setter also use `impl Into<#{nonOptionalType}>`?

Reply:
ah—I didn't do that because it makes editors get super confused and not show good hints.